rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/TextDocumentState.java

```diff
@@ -30,65 +30,78 @@
- * 
+ * contents to a tree. All other TextDocumentServices depend on this information.
+ *
  * Objects of this class are used by the implementations of RascalTextDocumentService
- * and ParametricTextDocumentService. 
+ * and ParametricTextDocumentService.
  */
 public class TextDocumentState {
     private final BiFunction<ISourceLocation, String, CompletableFuture<ITree>> parser;
 
     private final ISourceLocation file;
-    private volatile String currentContent;
+    private volatile Versioned<String> currentContent;
     @SuppressWarnings("java:S3077") // we are use volatile correctly
-    private volatile @MonotonicNonNull ITree lastFullTree;
+    private volatile @MonotonicNonNull Versioned<ITree> lastFullTree;
     @SuppressWarnings("java:S3077") // we are use volatile correctly
-    private volatile CompletableFuture<ITree> currentTree;
+    private volatile CompletableFuture<Versioned<@Nullable ITree>> currentTree;
 
-    public TextDocumentState(BiFunction<ISourceLocation, String, CompletableFuture<ITree>> parser, ISourceLocation file, String content) {
+    public TextDocumentState(BiFunction<ISourceLocation, String, CompletableFuture<ITree>> parser, ISourceLocation file, int version, String content) {
         this.parser = parser;
         this.file = file;
-        this.currentContent = content;
-        currentTree = newContents(content);
+        this.currentContent = new Versioned<>(version, content);
+        this.currentTree = newContent(version, content);
     }
 
-    public CompletableFuture<ITree> update(String text) {
-        currentContent = text;
-        currentTree = newContents(text);
+    /**
+     * The current call of this method guarantees that, until the next call,
+     * each intermediate call of `getCurrentTreeAsync` returns (a future for) a
+     * *correct* versioned tree. This means that:
+     *   - the version of the tree is parameter `version`;
+     *   - the tree is produced by parsing parameter `content`.
+     *
+     * Thus, callers of `getCurrentTreeAsync` are guaranteed to obtain a
+     * consistent <version, tree> pair.
+     */
+    public CompletableFuture<Versioned<ITree>> update(int version, String content) {
+        currentContent = new Versioned<>(version, content);
+        currentTree = newContent(version, content);
         return currentTree;
     }
 
     @SuppressWarnings("java:S1181") // we want to catch all Java exceptions from the parser
-    private CompletableFuture<ITree> newContents(String contents) {
-        return parser.apply(file, contents)
-            .whenComplete((r, t) -> { 
-                if (r != null) { 
-                    lastFullTree = r; 
-                } 
+    private CompletableFuture<Versioned<ITree>> newContent(int version, String content) {
+        return parser.apply(file, content)
+            .thenApply(t -> new Versioned<ITree>(version, t))
+            .whenComplete((r, t) -> {
+                if (r.get() != null) {
```

Comment:
I think `r != null` was done since `whenComplete` can actually have a `null` result in case of an exception. so I don't think this is safe.